File: core/src/sub/world/blocks/logic/UnitSpawner.java
package sub.world.blocks.logic;

import arc.graphics.*;
import arc.graphics.g2d.*;
import arc.math.*;
import arc.scene.ui.layout.*;
import arc.util.*;
import arc.util.io.*;
import mindustry.content.*;
import mindustry.ctype.*;
import mindustry.game.*;
import mindustry.gen.*;
import mindustry.graphics.*;
import mindustry.type.*;
import mindustry.ui.*;
import mindustry.ui.dialogs.*;
import sub.world.*;

import static mindustry.Vars.*;

public class UnitSpawner extends Structure{
    
    public UnitSpawner(String name){
        super(name);
        
        configurable = true;
        update = true;
    }
    
    public class UnitSpawnerBuild extends Building{
        public int unitf = UnitTypes.dagger.id;
        public int teamf;
        
        public int unitX = (int) x;
        public int unitY = (int) y;
        
        public void placed(){
            super.placed();
            
            teamf = team.id;
            resetPos();
        }
        
        @Override
        public void draw(){
            super.draw();
            
            Draw.alpha(0.75f);
            Draw.mixcol(teamf().color.cpy().mul(1f + Mathf.absin(Time.time() / 2f, 1f, 0.035f)), 1f);
            Draw.rect(unitf().icon(Cicon.full), this.unitX, this.unitY);
            Draw.reset();
        }
        
        public void buildConfiguration(Table table){
            table.button(Icon.fileText, () -> {
                unitDialog();
            }).size(40f);
            table.button(Icon.add, () -> {
                unitf().spawn(teamf(), this.unitX, this.unitY);
            }).size(40f);
        }
        
        public void unitDialog(){
            BaseDialog dialog = new BaseDialog("@dialog.name.unit-config");
            Table cont = dialog.cont;
            
            cont.table(t -> {
                t.top().margin(6f);
                t.add("@dialog.title.select-unit").color(Pal.accent).growX().center();
                t.row();
                t.image().color(Pal.accent).height(3f).pad(4f).fillX();
            }).width(800).center().row();
            
            cont.pane(p -> {
                int ru = 0;
                UnitType[] units = content.units().toArray(UnitType.class);
                
                for(UnitType unit : units){
                    p.button(b -> {
                        b.left();
                        b.image(unit.icon(Cicon.medium)).padRight(2).size(40f);
                        b.add(unit.localizedName);
                    }, () -> {
                        unitf(unit.id);
                    }).margin(12f).fillX();
                    
                    if(++ru % 3 == 0) p.row();
                };
            }).width(800f).height(540f).top().center().row();
            
            cont.table(i -> {
                i.table(t -> {
                    t.button("@dialog.title.select-team", Icon.modeSurvival, () -> {
                        teamDialog();
                    }).width(220f).pad(4f).growY();

                    t.button("@dialog.title.select-position", Icon.grid, () -> {
                        posDialog();
                    }).width(220f).pad(4f).growY();
                });
            }).width(360f).bottom().center();
            
            dialog.addCloseButton();
            dialog.show();
        }
        
        public void teamDialog(){
            BaseDialog dialog = new BaseDialog("$@dialog.title.select-team");
            Table cont = dialog.cont;

            cont.table(t -> {
                t.top().margin(6f);
                t.add("@dialog.info.base-teams").color(Pal.accent).growX();
                t.row();
                t.image().color(Pal.accent).height(3).pad(4).fillX();
            }).width(320f).center().row();
            
            cont.pane(p -> {
                int rt = 0;
                Team[] teams = Team.baseTeams;
                
                for(Team team : teams){                    
                    addTeamButton(p, team);
                    if(++rt % 3 == 0) p.row();
                }
            }).width(320f).center().row();
            
            cont.table(t -> {
                t.top().margin(6);
                t.add("@dialog.info.all-teams").color(Pal.accent).growX();
                t.row();
                t.image().color(Pal.accent).height(3).pad(4).fillX();
            }).width(320f).center().row();
            
            cont.pane(p -> {
                int rt = 0;
                Team[] teams = Team.all;
                
                for(Team team : teams){
                    addTeamButton(p, team);
                    if(++rt % 3 == 0) p.row();
                }
            }).width(320f).height(220f).center().row();
            
            cont.table(t -> {
                t.top().margin(6f);
                t.add("@dialog.short.others").color(Pal.accent).growX();
                t.row();
                t.image().color(Pal.accent).height(3).pad(4).fillX();
            }).width(320f).center().row();
            
            cont.table(t -> {
                t.button("@dialog.info.reset-team", () -> {
                    teamf(team.id);
                }).height(54f).pad(4f).growX().row();
                
                t.button("@dialog.info.set-team-id", () -> {
                    ui.showTextInput("", "@dialog.short.set-id", 8, String.valueOf(teamf().id), true, t2 -> {
                        int team = Integer.valueOf(t2);
                        
                        if(Team.get(team) != null){
                            teamf(team);
                        }else{
                            ui.showErrorMessage("@dialog.short.invalid-id");
                        }
                    });
                }).height(54f).pad(4f).growX();
            }).width(300f);
            
            dialog.addCloseButton();
            dialog.show();
        }
        
        public void posDialog(){
            BaseDialog dialog = new BaseDialog("@dialog.title.select-position");
            Table cont = dialog.cont;
            
            cont.table(t -> {
                t.button("@dialog.info.reset-position", () -> {
                    resetPos();
                }).height(54f).pad(4f).growX();
            }).width(300f).center().row();
            
            cont.table(t -> {
                t.top().margin(6f);
                t.add("@dialog.info.custom-position").color(Pal.accent).growX();
                t.row();
                t.image().color(Pal.accent).height(3).pad(4).fillX();
            }).width(320f).center().row();
            
            cont.table(t -> {
                int worldX = world.width();
                int worldY = world.height();
                
                t.button("@dialog.short.set-x", () -> {
                    ui.showTextInput("", "X:", 4, String.valueOf((this.unitX / tilesize)), true, x -> {
                        int tx = Integer.parseInt(x);
                    
                        if(tx <= worldX){
                            unitX(tx * tilesize);
                        }else{
                            ui.showInfo("@dialog.error.invalid-pos");
                        }
                    });
                }).growX().height(54f).pad(4f).row();
                
                t.button("@dialog.short.set-y", () -> {
                    ui.showTextInput("", "Y:", 4, String.valueOf((this.unitY / tilesize)), true, y -> {
                        int ty = Integer.parseInt(y);
                        
                        if(ty <= worldY){
                            unitY(ty * tilesize);
                        }else{
                            ui.showInfo("@dialog.error.invalid-pos");
                        }
                    });
                }).height(54f).pad(4f).growX();
            }).width(300f).center();
            
            dialog.addCloseButton();
            dialog.show();
        }
        
        public void addTeamButton(Table p, Team team){
            p.button(b -> {
                b.left();
                b.image().color(team.color).size(40f).pad(2f);
            }, () -> {
                teamf(team.id);
            }).pad(2f);
        }
        
        public void resetPos(){
            unitX((int) x);
            unitY((int) y);
        }
        
        public void write(Writes write){
            super.write(write);
            
            write.i(this.unitf);
            write.i(this.teamf);
            
            write.i(this.unitX);
            write.i(this.unitY);
        }
        
        public void read(Reads read){
            super.read(read);
            
            this.unitf = read.i();
            this.teamf = read.i();
            
            this.unitX = read.i();
            this.unitY = read.i();
        }
        
        public UnitType unitf(){
            return content.getByID(ContentType.unit, this.unitf);
        }
        
        public void unitf(int id){
            this.unitf = id;
        }
        
        public Team teamf(){
            return Team.get(this.teamf);
        }
        
        public void teamf(int id){
            this.teamf = id;
        }
        
        public int unitX(){
            return this.unitX;
        }
        
        public void unitX(int pos){
            this.unitX = pos;
        }
        
        public int unitY(){
            return this.unitY;
        }
        
        public void unitY(int pos){
            this.unitY = pos;
        }
    }
}

File: scripts/structures/logic/unitspawner.js
/*
    Copyright (C) 2020 Gdeft

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
const extendf = global.substructure.func.extendf;

const unitSpawner = extendf(Block, "unit-spawner", {
    requirements: ItemStack.with(Items.copper, 2, Items.lead, 4),
    buildVisibility: BuildVisibility.sandboxOnly,
    category: Category.logic,
    
    health: 40,
    size: 1,
    configurable: true,
    solid: false,

    load(){
        this.super$load();
        
        this.baseRegion = Core.atlas.find("substructure-base");
    },
    
    icons(){
        return [this.baseRegion, this.region];
    }
}, Building, {
    init(tile, team, shouldAdd, rotation){
        if(!this.initialized){
            this.create(tile.block(), team);
        }else{
            //Useless, but I don't want to remove it.
            if(this.block.hasPower){
                this.power.graph = new PowerGraph();
                this.power.graph.add();
            };
        };
        this.rotation = rotation;
        this.tile = tile;
        
        this.setUnit(UnitTypes.dagger.id);
        this.setTeam(team.id);
        
        this.set(tile.drawx(), tile.drawy());
        if(shouldAdd) this.add();
        
        this.created();
        return this;
        //print("Default Unit: " + this.getUnit() + "; Team: " + this.getTeam());
    },

    placed(){
        this.super$placed();

        this._unitX = this.getX();
        this._unitY = this.getY();
        //print("X: " + this.getUnitX() + " Y: " + this.getUnitY());
    },

    draw(){
        Draw.rect(this.blockf().baseRegion, this.x, this.y);
        
        Draw.alpha(0.75);
        Draw.mixcol(this.getTeam().color.cpy().mul(1 + Mathf.absin(Time.time / 2, 1, 0.035)), 1);
        Draw.rect(this.getUnit().icon(Cicon.full), this._unitX, this._unitY);
        Draw.rect(this.block.region, this.x, this.y);
        
        Draw.reset();
    },

    buildConfiguration(table){
        table.button(Icon.fileText, () => {
            this.unitDialog();
        }).size(40);
        table.button(Icon.add, () => {
            this.spawnUnit(this.getUnit(), this.getTeam(), this._unitX, this._unitY);
        }).size(40);
    },

    unitDialog(){
        let dialog = new BaseDialog("$dialog.name.unit-config");
        let cont = dialog.cont;
			
        cont.table(cons(t => {
            t.top().margin(6);
            t.add("$dialog.title.select-unit").growX().center().color(Pal.accent);
            t.row();
            t.image().fillX().height(3).pad(4).color(Pal.accent);
        })).width(800).center().row();

        cont.pane(cons(p => {
            let ru = 0;
            let units = Vars.content.units();
				
            units.each(cons(u => {
                p.button(cons(b => {
                    b.left();
                    b.image(u.icon(Cicon.medium)).scaling(Scaling.fit).size(40).padRight(6);
                    b.add(u.localizedName);
                }), () => {
                    this.setUnit(u.id);
                }).margin(12).fillX();

                if(++ru % 3 == 0) p.row();
            }));
        })).width(800).height(540).top().center().row();

        cont.table(cons(i => {
            i.table(cons(t => {
                t.button("$dialog.title.select-team", Icon.modeSurvival, () => {
                    this.teamDialog();
                }).width(220).pad(4).growY();

                t.button("$dialog.title.select-position", Icon.grid, () => {
                    this.posDialog();
                }).width(220).pad(4).growY();
            }));
        })).width(360).bottom().center();

        dialog.addCloseButton();
        dialog.show();
    },
		
    teamDialog(){
        let dialog = new BaseDialog("$dialog.title.select-team");
        let cont = dialog.cont;

        cont.table(cons(t => {
            t.top().margin(6);
            t.add("$dialog.info.base-teams").growX().color(Pal.accent);
            t.row();
            t.image().fillX().height(3).pad(4).color(Pal.accent);
        })).width(320).center().row();

        cont.pane(cons(p => {
            let rt = 0;
            let teams = Team.baseTeams;
				
            for(let i in teams){
                let team = teams[i];

                this.addTeamButton(p, team);
                if(++rt % 3 == 0) p.row();
            };
        })).width(320).center().row();

        cont.table(cons(t => {
            t.top().margin(6);
            t.add("$dialog.info.all-teams").growX().color(Pal.accent);
            t.row();
            t.image().fillX().height(3).pad(4).color(Pal.accent);
        })).width(320).center().row();

        cont.pane(cons(p => {
            let rt = 0;
            let teams = Team.all;

            for(let i in teams){
                let team = teams[i];
                
                this.addTeamButton(p, team);
                if(++rt % 3 == 0) p.row();
            };
        })).width(320).height(220).center().row();

        cont.table(cons(t => {
            t.top().margin(6);
            t.add("$dialog.short.others").growX().color(Pal.accent);
            t.row();
            t.image().fillX().height(3).pad(4).color(Pal.accent);
        })).width(320).center().row();

        cont.table(cons(t => {
            t.button("$dialog.info.reset-team", () => {
                this.setTeam(this.team.id);
            }).growX().height(54).pad(4).row();
				
            t.button("$dialog.info.set-team-id", () => {
                Vars.ui.showTextInput("", "$dialog.short.set-id", 8, this.getTeam().id, true, cons(t => {
                    if(Team.get(t) != null){
                        this.setTeam(t);
                    }else{
                        Vars.ui.showErrorMessage("$dialog.short.invalid-id");
                    };
                }));
            }).growX().height(54).pad(4);
        })).width(300);

        dialog.addCloseButton();
        dialog.show();
    },
		
    posDialog(){
        let dialog = new BaseDialog("$dialog.title.select-position");
        let cont = dialog.cont;

        cont.table(cons( t => {
            t.button("$dialog.info.reset-position", () => {
                this._unitX = this.getX();
                this._unitY = this.getY();
            }).growX().height(54).pad(4);
        })).width(300).center().row();

        cont.table(cons(t => {
            t.top().margin(6);
            t.add("$dialog.info.custom-position").growX().color(Pal.accent);
            t.row();
            t.image().fillX().height(3).pad(4).color(Pal.accent);
        })).width(320).center().row();

        cont.table(cons(t => {
            let worldX = Vars.world.width();
            let worldY = Vars.world.height();
				
            t.button("$dialog.short.set-x", () => {
                Vars.ui.showTextInput("", "X:", 4, (this._unitX / Vars.tilesize), true, cons(x => {
                    if(x <= worldX){
                        this._unitX = x * Vars.tilesize;
                    }else{
                        Vars.ui.showInfo("$dialog.error.invalid-pos");
                    };
                }));
            }).growX().height(54).pad(4).row();
				
            t.button("$dialog.short.set-y", () => {
                Vars.ui.showTextInput("", "Y:", 4, (this._unitY / Vars.tilesize), true, cons(y => {
                    if(y <= worldY){
                        this._unitY = y * Vars.tilesize;
                    }else{
                        Vars.ui.showInfo("$dialog.error.invalid-pos");
                    };
                }));
            }).growX().height(54).pad(4);
        })).width(300).center();

        dialog.addCloseButton();
        dialog.show();
    },

    addTeamButton(p, team){
        p.button(cons(b => {
            b.left();
            b.image().size(40).pad(2).color(team.color);
        }), () => {
            this.setTeam(team.id);
        }).pad(2);
    },

    spawnUnit(unit, team, x, y){ //There is no ${}
        let teamColor = "[#" + team.color + "]";

        unit.spawn(team, x, y);
        this.configure("Created unit " + teamColor + unit.localizedName + "[] at " + (x / Vars.tilesize) + ", " + (y / Vars.tilesize));
    },

    write(write){
        this.super$write(write);
	
        write.s(this._unit);
        write.s(this._team);

        write.s(this._unitX);
        write.s(this._unitY);
    },

    read(read, revision){ 
        this.super$read(read, revision);

        this._unit = read.s();
        this._team = read.s();

        this._unitX = read.s();
        this._unitY = read.s();
    },

    setUnit(id){
        this._unit = id;
    },

    getUnit(){
        return Vars.content.getByID(ContentType.unit, this._unit);
    },

    setTeam(id){
        this._team = id;
    },

    getTeam(){
        return Team.get(this._team);
	}
});